Add plumbing for loading web extensions in WebDriver

kiaraarose opened the pull request:
#### f5a4740781f10aa763ca8a05def0a18bad18aeb3
<pre>
Add plumbing for loading web extensions in WebDriver
<a href="https://bugs.webkit.org/show_bug.cgi?id=284393">https://bugs.webkit.org/show_bug.cgi?id=284393</a>
<a href="https://rdar.apple.com/138045738">rdar://138045738</a>

Reviewed by Timothy Hatcher and BJ Burg.

This patch adds the plumbing for loading web extensions in WebDriver.
See section Load WebExtension in the index.html file attached to this bug for the proposed
specification for this feature.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
Add feature flag.

* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::extractTemporaryZipArchive):
Move fix for <a href="https://rdar.apple.com/141179462">rdar://141179462</a> into this method so that this method returns the path to the subdirectory
containing the unzipped archive.

* Source/WebKit/UIProcess/API/APIAutomationSessionClient.h:
(API::AutomationSessionClient::loadWebExtensionWithOptions):

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKAutomationSessionDelegate.h:
Add a delegate method to handle loading the web extension in the browser.

* Source/WebKit/UIProcess/Automation/Automation.json:
Define the loadWebExtension method.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::loadWebExtension):
Parse the resource options and refer to the automation session client to load the web extension.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.mm:
(WebKit::AutomationSessionClient::AutomationSessionClient):
(WebKit::toAPI):
(WebKit::AutomationSessionClient::loadWebExtensionWithOptions):
Delegate loading the web extension to the browser.

Canonical link: <a href="https://commits.webkit.org/287737@main">https://commits.webkit.org/287737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6208657c3c6a995e49dd7b0d7a1c2e9316f15b36

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80591 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34524 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85115 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31573 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68659 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7904 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62954 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20756 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83660 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/53073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73370 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43259 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50366 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27530 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30032 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/73571 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71512 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28053 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86548 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/79650 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7817 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5516 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71247 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7992 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69207 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70487 "Found 2 new API test failures: /TestWebKit:WebKit2UserMessageRoundTripTest.WKURL, /TestWebKit:WebKit2TextFieldBeginAndEditEditingTest.TextFieldDidBeginShouldNotBeDispatchedForAlreadyFocusedField (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14497 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13447 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102057 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12498 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7779 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13298 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24847 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7618 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11137 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9423 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -255,6 +255,8 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     const String& displayDescription();
     const String& version();
 
+    URL resourceBaseURL();
```
```suggestion
    const URL& resourceBaseURL() const { return m_resourceBaseURL; }
```

elijahsawyers replied:
Ditto: https://github.com/WebKit/WebKit/pull/37729/files#r1879148684

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -747,6 +747,11 @@ const String& WebExtension::version()
+{
+    return m_resourceBaseURL;
+}
+
```
```suggestion
```

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
```diff
@@ -96,6 +96,8 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
 /*! @abstract A Boolean value indicating whether the extension has a sidebar. */
 @property (readonly, nonatomic) BOOL _hasSidebar;
 
+@property (readonly, nonatomic, copy) NSURL *_resourceURL;
```
You shouldn't need to add this. See Safari PR for details.

elijahsawyers commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
Isn't this change going away since we shouldn't need it anymore on the Safari-side?

elijahsawyers commented on Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
```diff
@@ -96,6 +96,8 @@ WK_HEADER_AUDIT_BEGIN(nullability, sendability)
 /*! @abstract A Boolean value indicating whether the extension has a sidebar. */
 @property (readonly, nonatomic) BOOL _hasSidebar;
 
+@property (readonly, nonatomic, copy) NSURL *_resourceURL;
```
Ditto: https://github.com/WebKit/WebKit/pull/37729/files#r1879148684

kiaraarose replied:
yes, but this PR went up before that discussion. just updated the PR

elijahsawyers commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -747,6 +747,11 @@ const String& WebExtension::version()
     return m_version;
 }
 
+URL WebExtension::resourceBaseURL()
```
Ditto: https://github.com/WebKit/WebKit/pull/37729/files#r1879148684

xeenon commented on Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
```diff
@@ -117,6 +117,14 @@ String extractTemporaryZipArchive(const String& path)
         BOMCopierFree(copier);
     }];
 
+    auto *contentsOfTemporaryDirectory = [NSFileManager.defaultManager contentsOfDirectoryAtPath:temporaryDirectory error:nil];
+    if (contentsOfTemporaryDirectory.count == 1) {
+        auto *subdirectoryPath = [temporaryDirectory stringByAppendingPathComponent:contentsOfTemporaryDirectory.firstObject];
+        BOOL isDirectory;
+        if ([NSFileManager.defaultManager fileExistsAtPath:subdirectoryPath isDirectory:&isDirectory] && isDirectory)
+            temporaryDirectory = subdirectoryPath;
+    }
```
You should remove the same code from `WebExtensionCocoa.mm`, as this handles it now.

kiaraarose replied:
oops, thought i did! must've left that change out. thanks for the catch

kiaraarose committed:
Add plumbing for loading web extensions in WebDriver
https://bugs.webkit.org/show_bug.cgi?id=284393
rdar://138045738

Reviewed by Timothy Hatcher and BJ Burg.

This patch adds the plumbing for loading web extensions in WebDriver.
See section Load WebExtension in the index.html file attached to this bug for the proposed
specification for this feature.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
Add feature flag.

* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::extractTemporaryZipArchive):
Move fix for rdar://141179462 into this method so that this method returns the path to the subdirectory
containing the unzipped archive.

* Source/WebKit/UIProcess/API/APIAutomationSessionClient.h:
(API::AutomationSessionClient::loadWebExtensionWithOptions):

* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKAutomationSessionDelegate.h:
Add a delegate method to handle loading the web extension in the browser.

* Source/WebKit/UIProcess/Automation/Automation.json:
Define the loadWebExtension method.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::loadWebExtension):
Parse the resource options and refer to the automation session client to load the web extension.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.h:
* Source/WebKit/UIProcess/Cocoa/AutomationSessionClient.mm:
(WebKit::AutomationSessionClient::AutomationSessionClient):
(WebKit::toAPI):
(WebKit::AutomationSessionClient::loadWebExtensionWithOptions):
Delegate loading the web extension to the browser.

Canonical link: https://commits.webkit.org/287737@main
